libzmix/src/signatures/delg_cred_cdd/groth_sig.rs

```diff
@@ -0,0 +1,439 @@
+use super::errors::{DelgError, DelgResult};
+use amcl_wrapper::extension_field_gt::GT;
+use amcl_wrapper::field_elem::{FieldElement, FieldElementVector};
+use amcl_wrapper::group_elem::{GroupElement, GroupElementVector};
+use amcl_wrapper::group_elem_g1::{G1LookupTable, G1Vector, G1};
+use amcl_wrapper::group_elem_g2::{G2Vector, G2};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct GrothSigkey(pub FieldElement);
+
+macro_rules! impl_GrothS {
+    ( $GrothSetupParams:ident, $GrothVerkey:ident, $GrothSig:ident, $GrothS:ident, $vk_group:ident, $msg_group:ident, $GVector:ident ) => {
+        #[derive(Clone, Debug, Serialize, Deserialize)]
+        pub struct $GrothSetupParams {
+            pub g1: G1,
+            pub g2: G2,
+            pub y: $GVector,
+        }
+
+        #[derive(Clone, Debug, Serialize, Deserialize)]
+        pub struct $GrothVerkey(pub $vk_group);
+
+        #[derive(Clone, Debug, Serialize, Deserialize)]
+        pub struct $GrothSig {
+            pub R: $vk_group,
+            pub S: $msg_group,
+            pub T: $GVector,
+        }
+
+        pub struct $GrothS {}
+    };
+}
+
+macro_rules! impl_GrothS_setup {
+    ( $GrothSetupParams:ident, $msg_group:ident, $GVector:ident ) => {
+        pub fn setup(count_messages: usize, label: &[u8]) -> $GrothSetupParams {
+            // NUMS for g1 and g2
+            let g1 = G1::from_msg_hash(&[label, " : g1".as_bytes()].concat());
+            let g2 = G2::from_msg_hash(&[label, " : g2".as_bytes()].concat());
+            let mut y = $GVector::with_capacity(count_messages);
+            for i in 0..count_messages {
+                // // NUMS for y. construct a group element from hashing label||y||i for each i
+                let yi = $msg_group::from_msg_hash(
+                    &[label, " : y".as_bytes(), i.to_string().as_bytes()].concat(),
+                );
+                y.push(yi);
+            }
+            $GrothSetupParams { g1, g2, y }
+        }
+    };
+}
+
+macro_rules! impl_GrothSig_new {
+    ( $messages:ident, $sk:ident, $y:expr, $g_r:expr, $g_s:expr, $msg_group_vec:ident ) => {{
+        if $messages.len() > $y.len() {
+            return Err(DelgError::UnsupportedNoOfMessages {
+                expected: $y.len(),
+                given: $messages.len(),
+            });
+        }
+        let r = FieldElement::random();
+        let r_inv = r.inverse();
+        let R = &$g_r * &r;
+        let S = (&$y[0] + (&$g_s * &$sk.0)) * &r_inv;
+        let mut T = $msg_group_vec::with_capacity($messages.len());
+        for i in 0..$messages.len() {
+            T.push(&$messages[i] + (&$y[i] * &$sk.0));
+        }
+        T.scale(&r_inv);
+        Ok(Self { R, S, T })
+    }};
+}
+
+macro_rules! impl_GrothSig_randomize {
+    (  ) => {
+        pub fn randomize(&self, r_prime: &FieldElement) -> Self {
```

Comment:
Is `randomize` the correct name for this? It's more like a special type of multiply by a FieldElement.

Reply:
This multiplication randomizes the signature.

Reply:
Added above comment as code comment